rosbag2_storage/include/rosbag2_storage/storage_interfaces/base_read_interface.hpp

```diff
@@ -25,6 +25,27 @@
 
 namespace rosbag2_storage
 {
+
+struct ReadOrder
+{
+  enum SortBy
+  {
+    ReceivedTimestamp,
+    PublishedTimestamp,  // NOTE: not yet implemented in ROS 2 core
+    File
+  };
+
+  // Sorting criterion for reading out messages, ascending by default
+  SortBy sort_by = ReceivedTimestamp;
+  // If true, changes sort order to descending
+  bool reverse = false;
```

Comment:
I disagree with the base premise. This is not an class object with behaviors - it's a POD struct only intended to carry a bundle of configuration values. All of this class behavior and extra code is absolutely unnecessary complexity - `struct`s exist for a reason in the language and are frequently used in this way.

Reply:
I would 100% agree if it would be really a POD struct or POD type.
```
Value of: std::is_pod<rosbag2_storage::ReadOrder>()
  Actual: false
Expected: true
```
It has comparison operator and the way how we are going to use this type looks like it needs constructor with two parameters. Almost each time we will have to initialize both member variables after creation of `ReaderOrder` instance.  
For instance:
https://github.com/ros2/rosbag2/blob/167182a2718f8a327470e72b35e383d64044f8a5/rosbag2_cpp/test/rosbag2_cpp/test_sequential_reader.cpp#L301-L303

I agree that it's tiny class with just two member variables and it's convenient to have direct access to them, although we have coding style guidance for such case 
> - we should have a good reason for allowing direct member access, other than because it is convenient for us

However even if I have my personal preferences to see it implemented as class I don't have a strict request for this changes and let you as an author make a final decision by yourself.

Reply:
I realize that at some point this comes down to a matter of taste - but this pattern is also backed up by the style guide. [C++ section of ROS style guide](https://docs.ros.org/en/rolling/The-ROS2-Project/Contributing/Code-Style-Language-Versions.html#id3) says we're using the [Google C++ Style Guide] (https://google.github.io/styleguide/cppguide.html) with the stated modifications. The section on [structs vs classes](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes) says
> Use a struct only for passive objects that carry data; everything else is a class.
> structs should be used for passive objects that carry data, and may have associated constants. All fields must be public. The struct must not have invariants that imply relationships between different fields, since direct user access to those fields may break those invariants. Constructors, destructors, and helper methods may be present; however, these methods must not require or enforce any invariants.

Which is exactly the situation of this type of data-carrying record object `ReadOrder`. Though it doesn't fit the strict `std::is_pod` check (only because of the default values on the members, without those it would pass), it is still "Plain Old Data" in the general sense, and in the sense of the style guide.

The section about Access Control that you pointed to was intended to be _less strict_ than the Google guide access control for things that are `class`es rather than `struct`s from the above criteria. The part you quoted comes as a sub-point of 
> Drop requirement for all class members to be private and therefore require accessors

The wording is intended to be a "consider" and "prefer" wording, rahter than "require", which the Google guide does require. However, it doesn't apply to this case anyways.
 
Beyond the style guide, which is a guide and not a law, I would also argue for consistency with the existing API pattern set by `StorageOptions` etc, rather than add a new pattern to the APIs.

I am going to plan to move forward with the struct the way it is. However, I will add a constructor for convenience.

Reply:
Thanks, constructor will be really helpful.